light-client/src/operations/voting_power.rs

```diff
@@ -1,19 +1,75 @@
 use crate::{
     bail,
     predicates::errors::VerificationError,
-    types::{SignedHeader, ValidatorSet},
+    types::{Commit, SignedHeader, TrustThreshold, ValidatorSet},
 };
 
-use anomaly::BoxError;
+use serde::{Deserialize, Serialize};
+use std::collections::HashSet;
+use std::fmt;
+
+use tendermint::block::CommitSig;
+use tendermint::lite::types::TrustThreshold as _;
 use tendermint::lite::types::ValidatorSet as _;
+use tendermint::vote::{SignedVote, Vote};
+
+#[derive(Copy, Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct VotingPower {
+    pub total: u64,
+    pub tallied: u64,
+    pub trust_threshold: TrustThreshold,
+}
+
+impl fmt::Display for VotingPower {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "VotingPower(total={} tallied={} trust_threshold={})",
+            self.total, self.tallied, self.trust_threshold
+        )
+    }
+}
 
 pub trait VotingPowerCalculator: Send {
     fn total_power_of(&self, validators: &ValidatorSet) -> u64;
-    fn voting_power_in(
+
+    fn check_enough_trust(
+        &self,
+        untrusted_header: &SignedHeader,
+        untrusted_validators: &ValidatorSet,
+        trust_threshold: TrustThreshold,
+    ) -> Result<VotingPower, VerificationError> {
+        let voting_power =
+            self.voting_power_of(untrusted_header, untrusted_validators, trust_threshold)?;
+
+        if trust_threshold.is_enough_power(voting_power.tallied, voting_power.total) {
```

Comment:
Mentioned in one of the comments above, I would probably make more sense to just check this in the `voting_power_of` and instead of a value just return a boolean value or something else to tell us if everything is fine or there is some error.